submit_cgap/submission.py

```diff
@@ -0,0 +1,354 @@
+
+        tracking_url = ingestion_submission_item_url(server=server, uuid=uuid)
+
+        outcome = None
+        n_tries = 8
+        tries_left = n_tries
+        done = False
+        while tries_left > 0:
+            # Pointless to hit the queue immediately, so we avoid some
+            # server stress by sleeping even before the first try.
+            time.sleep(PROGRESS_CHECK_INTERVAL)
+            res = res = requests.get(tracking_url, auth=auth_tuple).json()
+            processing_status = res['processing_status']
+            done = processing_status['state'] == 'done'
+            if done:
+                outcome = processing_status['outcome']
+                break
+            else:
+                progress = processing_status['progress']
+                show("Progress is %s. Continuing to wait..." % progress, with_time=True)
+            tries_left -= 1
+
+        if not done:
+            show("Timed out after %d tries." % n_tries, with_time=True)
+            exit(1)
+
+        show("Final status: %s" % outcome, with_time=True)
+
+        if outcome == 'error' and res.get('errors'):
+            show_section(res, 'errors')
+
+        caveat_outcome = None if outcome == 'success' else outcome
+
+        show_section(res, 'validation_output', caveat_outcome=caveat_outcome)
+
+        if validate_only:
+            exit(0)
+
+        show_section(res, 'post_output', caveat_outcome=caveat_outcome)
+
+        if outcome == 'success':
+            show_section(res, 'upload_info')
+            do_any_uploads(res, auth_dict, bundle_folder=None, bundle_filename=bundle_filename)
+
+
+def do_any_uploads(res, auth_dict, bundle_folder=None, bundle_filename=None):
+    upload_info = get_section(res, 'upload_info')
+    if upload_info:
+        if yes_or_no("Upload %s?" % n_of(len(upload_info), "file")):
+            do_uploads(upload_info, auth=auth_dict,
+                       folder=bundle_folder or (os.path.dirname(bundle_filename) if bundle_filename else None))
+
+
+def resume_uploads(uuid, server=None, env=None, bundle_filename=None, auth_dict=None):
+
+    with script_catch_errors():
+
+        server = resolve_server(server=server, env=env)
+        auth_dict = auth_dict or get_cgap_auth_dict(server)
+        url = ingestion_submission_item_url(server, uuid)
+        response = requests.get(url, auth=cgap_auth_to_tuple(auth_dict))
+        response.raise_for_status()
+        do_any_uploads(response.json(), auth_dict, bundle_filename=bundle_filename or os.path.abspath(os.path.curdir))
+
+
+def execute_prearranged_upload(path, upload_credentials):
+    """
+    This performs a file upload using special credentials received from ff_utils.patch_metadata.
+
+    :param path: the name of a local file to upload
+    :param upload_credentials: a dictionary of credentials to be used for the upload,
+        containing the keys 'AccessKeyId', 'SecretAccessKey', 'SessionToken', and 'upload_url'.
```

Comment:
Yeah, let's try. It's the one PyCharm seems to insert if you write `"""` (and it inserts `"""` after the cursor leaving you between) and then press `Enter`. It will autofill the parameters, and it will whine that it's a PEP8 violation if some of the arguments are missing. So yes, I guess I'm OK with it.